homeassistant/components/websocket_api/messages.py

```diff
@@ -74,6 +74,110 @@ def _cached_event_message(event: Event) -> str:
     return message_to_json(event_message(IDEN_TEMPLATE, event))
 
 
+def cached_state_diff_message(iden: int, event: Event) -> str:
+    """Return an event message.
+
+    Serialize to json once per message.
+
+    Since we can have many clients connected that are
+    all getting many of the same events (mostly state changed)
+    we can avoid serializing the same data for each connection.
+    """
+    return _cached_state_diff_message(event).replace(IDEN_JSON_TEMPLATE, str(iden), 1)
+
+
+@lru_cache(maxsize=128)
+def _cached_state_diff_message(event: Event) -> str:
+    """Cache and serialize the event to json.
+
+    The IDEN_TEMPLATE is used which will be replaced
+    with the actual iden in cached_event_message
+    """
+    return message_to_json(event_message(IDEN_TEMPLATE, _state_diff_event(event)))
+
+
+def _state_diff_event(event: Event) -> dict:
+    """Convert a state_changed event to the minimal version.
+
+    State update example
+
+    {
+        add: {entity_id: state,…}
+        changed: {entity_id: diff,…}
+        removed: [entity_id,…]
+    }
```

Comment:
We could change the payload type from 'add' => 'a', 'remove' => 'r', 'changed' => 'c' ?

Reply:
Since its a new api I went with it